devices/src/pvmemcontrol.rs

```diff
@@ -0,0 +1,819 @@
+// Copyright © 2024 Google LLC
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use num_enum::TryFromPrimitive;
+use pci::{
+    BarReprogrammingParams, PciBarConfiguration, PciBarPrefetchable, PciBarRegionType,
+    PciClassCode, PciConfiguration, PciDevice, PciDeviceError, PciHeaderType, PciSubclass,
+};
+use std::{
+    collections::HashMap,
+    ffi::CString,
+    io, result,
+    sync::{Arc, Barrier, Mutex, RwLock},
+};
+use thiserror::Error;
+use vm_allocator::{page_size::get_page_size, AddressAllocator, SystemAllocator};
+use vm_device::{BusDeviceSync, Resource};
+use vm_memory::{
+    bitmap::AtomicBitmap, Address, ByteValued, Bytes, GuestAddress, GuestAddressSpace, GuestMemory,
+    GuestMemoryAtomic, GuestMemoryError, GuestMemoryMmap, Le32, Le64,
+};
+use vm_migration::{Migratable, MigratableError, Pausable, Snapshot, Snapshottable, Transportable};
+
+const PVMEMCONTROL_VENDOR_ID: u16 = 0x1ae0;
+const PVMEMCONTROL_DEVICE_ID: u16 = 0x0087;
+
+const PVMEMCONTROL_SUBSYSTEM_VENDOR_ID: u16 = 0x1ae0;
+const PVMEMCONTROL_SUBSYSTEM_ID: u16 = 0x011F;
+
+const MAJOR_VERSION: u64 = 1;
+const MINOR_VERSION: u64 = 0;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    // device errors
+    #[error("Guest gave us bad memory addresses: {0}")]
+    GuestMemory(#[source] GuestMemoryError),
+    #[error("Guest sent us invalid request")]
+    InvalidRequest,
+
+    #[error("Guest sent us invalid command")]
+    InvalidCommand(u32),
+    #[error("Guest sent us invalid connection")]
+    InvalidConnection(u32),
+
+    // pvmemcontrol errors
+    #[error("Request contains invalid arguments")]
+    InvalidArgument(u64),
+    #[error("Unknown function code: {0}")]
+    UnknownFunctionCode(u64),
+    #[error("Libc call fail.: {0}")]
+    LibcFail(#[source] std::io::Error),
+}
+
+#[derive(Copy, Clone)]
+enum PvmemcontrolSubclass {
+    Other = 0x80,
+}
+
+impl PciSubclass for PvmemcontrolSubclass {
+    fn get_register_value(&self) -> u8 {
+        *self as u8
+    }
+}
+
+/// commands have 0 as the most significant byte
+#[repr(u32)]
+#[derive(PartialEq, Eq, Copy, Clone, TryFromPrimitive)]
+enum PvmemcontrolTransportCommand {
+    Reset = 0x060f_e6d2,
+    Register = 0x0e35_9539,
+    Ready = 0x0ca8_d227,
+    Disconnect = 0x030f_5da0,
+    Ack = 0x03cf_5196,
+    Error = 0x01fb_a249,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct PvmemcontrolTransportRegister {
+    buf_phys_addr: Le64,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct PvmemcontrolTransportRegisterResponse {
+    #[allow(dead_code)]
```

Comment:
Why is this needed? I see that this structure is used below.

Reply:
Fixed